File: AlgorithmsPackCPP/785. Is Graph Bipartite.cpp
class Solution {
    bool colorConflict{};
public:
    void dfs(vector<vector<int>>&graph, int node, vector<int>&colors, int assignCol = 1){
        if(colors[node]==0)
            colors[node] = assignCol;
        else{
            if(colors[node]!=assignCol)
                colorConflict = true;

            return;
        }
        for(int cur:graph[node])
            dfs(graph,cur,colors,3-assignCol);
    }

    bool isBipartite(vector<vector<int>>& graph) {
        int n = graph.size();
        vector<int>colors(n,0);

        for(int i = 0; i < n; i++){
            if(colors[i]==0){
                dfs(graph,i,colors);
                if(colorConflict)
                    return false;
            }
        }
        return true;
    }
};

File: AlgorithmsPackPythonSolutions/785. Is Graph Bipartite.py
class Solution:
    def isBipartite(self, graph: List[List[int]]) -> bool:
        colorConflict = False
        
        def dfs(graph, node, colors, assignCol=1):
            if colors[node] == 0:
                colors[node] = assignCol
            else:
                if colors[node] != assignCol:
                    nonlocal colorConflict
                    colorConflict = True
                return
            
            for cur in graph[node]:
                dfs(graph, cur, colors, 3 - assignCol)
        
        n = len(graph)
        colors = [0] * n
        
        for i in range(n):
            if colors[i] == 0:
                dfs(graph, i, colors)
                if colorConflict:
                    return False
        
        return True
